### PR TITLE
GraphQL POC

### LINKED ISSUE #60: Enchancment: GraphQL Support
Honestly, this would be kind of cool to have.

### PR DESCRIPTION
**Pull request by bagelbits**
## What does this do?
Adds graphql support to the API

## How was it tested?
Locally with setting the realm credentials as ENV variables.

## Is there a Github issue this is resolving?
This resolves #60, finally.

## Here's a fun image for your troubles
image

### CODE REVIEW

File: src/apollo/server.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+const { ApolloServer } = require('apollo-server-express');
```
Comment (bagelbits) [inline src/apollo/server.js:1]: I think I might want to move this into the middleware folder for now. It doesn't make sense to have a dedicated Apollo folder right now.

File: src/apollo/server.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+const fetch = require('node-fetch');
+const { introspectSchema, wrapSchema } = require('graphql-tools');
+const { print } = require('graphql');
+const { graphqlUrl, getValidAccessToken, realmAvailable } = require('../util/RealmClient');
+
+const executor = async ({ document, variables, _context }) => {
```
Comment (bagelbits) [inline src/apollo/server.js:7]: The pattern can be found [here.](https://www.graphql-tools.com/docs/remote-schemas)

Comment (bagelbits) [reply src/apollo/server.js:7]: I might leave a comment in code so I remember where this came form.

File: src/util/RealmClient.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
+const Realm = require('realm-web');
+const { realmAppId, realmApiKey } = require('./environmentVariables');
+
+// Connect to your MongoDB Realm app
+const app = new Realm.App(realmAppId);
```
Comment (bagelbits) [inline src/util/RealmClient.js:5]: This pattern is pulled from [here.](https://docs.mongodb.com/realm/web/graphql-apollo-react/)

Comment (bagelbits) [reply src/util/RealmClient.js:5]: Can probably add a comment as well so I can find this again.

File: src/util/RealmClient.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+const Realm = require('realm-web');
+const { realmAppId, realmApiKey } = require('./environmentVariables');
+
+// Connect to your MongoDB Realm app
```
Comment (bagelbits) [inline src/util/RealmClient.js:4]:
```suggestion

// Pattern copied from https://docs.mongodb.com/realm/web/graphql-apollo-react/
// Connect to your MongoDB Realm app
```

File: src/apollo/server.js (hunk 2 above)
Comment (bagelbits) [inline src/apollo/server.js:7]:
```suggestion

// This pattern copied from https://www.graphql-tools.com/docs/remote-schemas/
const executor = async ({ document, variables, _context }) => {
```

Comment (bagelbits) [reply src/apollo/server.js:8]: Added.

File: src/util/RealmClient.js (hunk 3 above)
Comment (bagelbits) [reply src/util/RealmClient.js:6]: Added.

File: jest.config.integration.js [hunk 5]
```diff
@@ -1,4 +1,4 @@
 const config = require('./jest.config');
-config.testRegex = '.itest\\.js$'; //Overriding testRegex option
+config.testRegex = '\\.itest\\.js$'; //Overriding testRegex option
```
Comment (Redmega) [inline jest.config.integration.js:2]: Is it possible to just use an inline regex for this? `/\.itest\.js$/` might look _slightly_ cleaner

**Review by Redmega — APPROVED** [review]
Mainly nits, feel free to ignore

File: jest.config.integration.js (hunk 5 above)
Comment (bagelbits) [reply jest.config.integration.js:2]: Unfortunately, it is not. Jest wants a string or array of strings. It won't accept a raw regex.

File: src/middleware/apolloServer.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    schema: await introspectSchema(executor),
+    executor
+  });
+  const server = new ApolloServer({ schema });
+  return server;
+};
+
+const apolloAvailable = () => {
+  return realmAvailable();
+};
```
[the remark this answers is not in the document]
Comment (bagelbits) [reply src/middleware/apolloServer.js]: Done!

**Pull request merged**